src/Equations/Fluid/LCFSLossRateTerm.cpp

```diff
@@ -0,0 +1,208 @@
+    RunawaySourceTerm(grid, unknowns), FVM::DiagonalComplexTerm(grid, unknowns, operandGrid), unknowns(unknowns), scaleFactor(sf), t_loss(t_loss), 
+    userGivenPsiEdge_t0(userGivenPsiEdge_t0), psi_edge_t0(PsiEdge_t0), 
+    id_psi(unknowns->GetUnknownID(OptionConstants::UQTY_POL_FLUX)) {
+
+    SetName("LCFSLossRateTerm");
+    this->rGrid = grid->GetRadialGrid();
+    this->GridRebuilt();
+}
+
+/**
+ * Destructor
+ */
+LCFSLossRateTerm::~LCFSLossRateTerm(){
+    Deallocate();
+}           
+
+
+/**
+ * Called after the grid is rebuilt; (re)allocates memory
+ * for all quantities
+ */
+bool LCFSLossRateTerm::GridRebuilt(){
+
+    this->DiagonalComplexTerm::GridRebuilt();
+    Deallocate();
+    this->GammaLoss = new real_t[nr];
+    
+    return true;
+}
+
+
+/**
+* Set weights for term to be multiplied with unknown (n_RE). 
+*/
+void LCFSLossRateTerm::SetGammaLoss(){
+
+    for (len_t ir = 0; ir < nr; ir++){ // Empty GammaLoss
+        this->GammaLoss[ir] = 0;
+    }
+    
+    FindRadiusOfLCFS(); // Find ir_LCFS
+    
+    for (len_t ir = 0; ir < nr; ir++){
+        this->GammaLoss[ir] = -StepFunction(ir) / this->t_loss[ir]; // Change GammaLoss from 0 if ir>ir_LCFS
+    }
+}
+
+
+
+/**
+* Find current radial point of last closed flux surface
+* and find psi_edge_t0 if not provided in input
+*/
+void LCFSLossRateTerm::FindRadiusOfLCFS(){
+    
+    if(this->userGivenPsiEdge_t0 == 0) {
+        //this->psi_edge_t0 = unknowns->GetUnknownInitialData(id_psi)[nr-1];
+        
+        // Need special solution if nr=1(0)?
+        // Extrapolate psi from last two ir:s to find the approximate edge value at t=0
+        real_t Rlow = this->rGrid->GetR(nr-2);
+        real_t Rhigh = this->rGrid->GetR(nr-1);
+        real_t Psilow = unknowns->GetUnknownInitialData(id_psi)[nr-2];
+        real_t Psihigh = unknowns->GetUnknownInitialData(id_psi)[nr-1];
+        real_t slope = (Psihigh - Psilow) / (Rhigh - Rlow);
+        real_t Redge = this->rGrid->GetR_f(nr);
+        
+        this->psi_edge_t0 = Psihigh + slope * (Redge - Rhigh); 
+    }
+    
+    // Find if the sign of PsiDiff() should be +1 or -1
+    if(this->signFixed == false) { 
+        real_t psi_mid_init = unknowns->GetUnknownInitialData(id_psi)[0]; // Use interpolate instead?
+        real_t psi_edge_init = unknowns->GetUnknownInitialData(id_psi)[nr-1];
+        if(psi_mid_init-psi_edge_init > 0) { this->sign = 1; }
+        this->signFixed = true;
+    }
+    
+    bool exists = false;
+    for (len_t ir = 0; ir < nr; ir++){
+        if(this->sign * PsiDiff(ir) >= 0){
+            this->ir_LCFS = (int_t)ir; // Last ir to pass critera is ir_LCFS
+            exists = true;
+        }
+    }
+    if ( exists == false ) { this->ir_LCFS = -1; } // No flux surface is closed
+}
+
+
+
+/**
+* Difference between the poloidal flux at radial point ir and at the edge at t=0
```

Comment:
Doesn't this routine return the difference between ``psi_p`` at ``ir+1/2`` and the edge ``psi_p``?